Add steps to execute module tests in Travis CI and fix openstack module unit test

relaxdiego opened the pull request:
##### ISSUE TYPE
- Bugfix Pull Request
##### COMPONENT NAME

openstack and .travis.yml file
##### ANSIBLE VERSION

```
ansible 2.2.0 (devel e81f14ab48) last updated 2016/06/09 09:45:16 (GMT -700)
  lib/ansible/modules/core: (run_unit_tests e9fd11d328) last updated 2016/06/09 21:09:03 (GMT -700)
  lib/ansible/modules/extras: (detached HEAD 93b59ba852) last updated 2016/06/09 09:45:29 (GMT -700)
  config file = 
  configured module search path = Default w/o overrides
```
##### SUMMARY

This change executes the tests that come with the repo in Travis CI. It also fixes an error in the openstack unit test.

gregdek commented:
Thanks @relaxdiego. To the current maintainers, @emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

Shrews commented:
I didn't even know these OpenStack unit tests existed (we didn't create them, afaik). And honestly, I don't see what this is fixing in the existing unit test. The 'meta' param is totally optional and don't see what adding it is fixing. On the other hand, I don't see that adding it hurts anything either, so I won't object to the change.

bcoca commented:
@relaxdiego can you show an example of the error?

relaxdiego commented:
@bcoca yes. I encountered it while running the tests in Travis using an earlier commit from my fork: https://travis-ci.org/relaxdiego/ansible-modules-core/builds/136601920

relaxdiego commented:
Pasting the Travis CI error here too for convenience:

```
The command "pip install -r test-requirements.txt" exited with 0.
0.36s$ PYTHONPATH=$PYTHONPATH:. py.test --random test
============================= test session starts ==============================
platform linux2 -- Python 2.7.9 -- py-1.4.26 -- pytest-2.6.4
Tests are shuffled using seed number 375175617844.
plugins: random
collected 10 items / 1 errors 
test/unit/cloud/openstack/test_os_server.py ........F.
==================================== ERRORS ====================================
_____ ERROR collecting test/unit/cloud/somebodyscomputer/test_firstmod.py ______
test/unit/cloud/somebodyscomputer/test_firstmod.py:4: in <module>
    from ansible.modules.core.cloud.somebodyscomputer import firstmod
E   ImportError: No module named somebodyscomputer
=================================== FAILURES ===================================
_____________________ TestCreateServer.test_create_server ______________________
self = <test_os_server.TestCreateServer object at 0x7f1b630f5290>
    def test_create_server(self):
        '''
            - os_server:
                image: cirros
                flavor: m1.tiny
                nics:
                  - net-name: network1
            '''
        with pytest.raises(AnsibleExit):
>           os_server._create_server(self.module, self.cloud)
test/unit/cloud/openstack/test_os_server.py:185: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
module = <MagicMock id='139755602792912'>
cloud = <test_os_server.FakeCloud object at 0x7f1b630f52d0>
    def _create_server(module, cloud):
        flavor = module.params['flavor']
        flavor_ram = module.params['flavor_ram']
        flavor_include = module.params['flavor_include']

        image_id = None
        if not module.params['boot_volume']:
            image_id = cloud.get_image_id(
                module.params['image'], module.params['image_exclude'])

        if flavor:
            flavor_dict = cloud.get_flavor(flavor)
            if not flavor_dict:
                module.fail_json(msg="Could not find flavor %s" % flavor)
        else:
            flavor_dict = cloud.get_flavor_by_ram(flavor_ram, flavor_include)
            if not flavor_dict:
                module.fail_json(msg="Could not find any matching flavor")

        nics = _network_args(module, cloud)

        if type(module.params['meta']) is str:
            metas = {}
            for kv_str in module.params['meta'].split(","):
>               k, v = kv_str.split("=")
E               ValueError: need more than 1 value to unpack
cloud/openstack/os_server.py:444: ValueError
================= 1 failed, 9 passed, 1 error in 0.13 seconds ==================
The command "PYTHONPATH=$PYTHONPATH:. py.test --random test" exited with 1.
```

bcoca commented:
that looks like a bug in the test that assumes meta is always present.

relaxdiego commented:
Actually, that error is from the openstack os_server module code itself, not the test.

Shrews commented:
I think that works in os_server.py outside of that unit test because if meta is not given, it's type would be NoneType, not str. For some reason, within the unit test, if meta is not supplied, it's a str type, so it fails.

relaxdiego commented:
I see. I can modify the os_server code to check for NoneType and empty string.

Shrews commented:
@relaxdiego That was total speculation on my part. Might want to verify that it does get NoneType before proposing a change.  :)

relaxdiego commented:
I see the cause now. It's from this line from the method `params_from_doc` in the test: `yaml.load()`

It returns a defaultdic object where the default value is an empty string.

I tried changing `module.params['meta']` to default to None when not provided but that still results in a failure because now the `_create_server` code will try to call `split(',')` on a NoneType.

relaxdiego commented:
Correction on that last part: I was in trace mode inside the `if type(module.params['meta']) is str` conditional when I changed params['meta'] to None. I should have tried doing that before the loop. In that case, it will **not** encounter an error if params['meta'] is NoneType.

relaxdiego commented:
Well, that fixed the problem but my new test caused a new test error. :) Seems like self.cloud doesn't get reset between tests. Hmmm...

relaxdiego commented:
The latest test error was caused by `FakeCloud.create_server` being a class attribute. Thus, it's never re-initialized between instantiation of the class. I turned it into a property that uses an instance attribute that gets initialized with the object and that solved it.

relaxdiego commented:
Sorry, I'm so used to Gerrit I forgot that overwriting previous commits would also cause GitHub to remove them from this thread.

sivel commented:
cc @mattclay 

This seems to partially align with the v2.2 Roadmap:

> Core Module Test Organization
> Relocate core module tests to ansible-modules-core to encourage inclusion of tests in core module PRs

aimonb commented:
Anything that moves us towards having functional tests is a good thing IMO. +1

relaxdiego commented:
I will not indulge my OCD side anymore. I'm happy with that last commit if everyone else is. :)

Shrews commented:
Just as an aside for those interested, we have more complete functional testing of the OpenStack modules within the shade library itself:

  https://github.com/openstack-infra/shade/tree/master/shade/tests/ansible

We test each new shade patch against both the released version of Ansible and the current HEAD of devel. We eventually hope to make this more of a two-way street, but that is a bit down the road.

sivel commented:
Just to link it here, the tests were added as part of https://github.com/ansible/ansible-modules-core/pull/2286

Seemingly to test the fix in https://github.com/ansible/ansible-modules-core/pull/2275

Maybe they aren't even needed any more?

aimonb commented:
@sivel Tests are exactly what is needed! Fixes are never permanent unless you create tests to prove it. Regressions are common when no tests exist for a given functionality. Tests prevent future breakage! We need more tests not fewer.

sivel commented:
@aimonb I get that, but to date the lieutenants of the openstack modules have decided to and have been maintaining their tests outside of the module repo.

I agree we need more and better tests, but we should remain consistent and maintain the tests where the primary maintainers know and can maintain properly.

relaxdiego commented:
That's an interesting setup. IMHO, the decision to remove that should be done by someone from the openstack team and this PR should just be limited to running the tests and making sure pre-existing tests pass.

aimonb commented:
@sivel Thanks for the info. Interesting indeed @relaxdiego. The issue as I understand it is that without this fix the builds will always fail. That kind of defeats the pass/fail mechanism. So I agree we should not do any work to extend the OpenStack stuff since they are managing externally, but we do need this fix so that contributors and reviewers quickly know if new changes cause build to fail.

relaxdiego commented:
To add to that, our team is in the process of contributing a cloud module and would like to see feedback on our tests in Travis.

relaxdiego commented:
I modified the commit to use nose instead of pytest to conform with the rest of Ansible's test specs. Also, I purposely didn't add an **init**.py to the openstack unit tests @sivel suggested that it might not be needed anymore. (Nose doesn't run tests in a dir that doesn't have that file) I'll leave it to the maintainers of the openstack test to decide what to do with it.

relaxdiego committed:
Run unit tests in Travis CI

This change allows module developers to execute their tests
when they submit a PR. For this to happen, the unit test subdir
where the tests exist must be a package (must have __init__.py)
The openstack unit tests purposely do not have an __init__.py
because it's failing and it's not written to work with nosetests
which is the de facto testing framework being used by the Ansible.

gregdek commented:
Thanks @relaxdiego for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

relaxdiego commented:
Closing this PR since this is no longer necessary.
